### PR TITLE
Issue1 implementation

### PR DESCRIPTION
**Pull request by golenkovm**
https://github.com/catalyst/moodle-tool_mfa/issues/1

- Show grace period info page if there is no configured factor for given user.
- Use get_enabled_user_factors() method for simplification.
- Authenticate factors one by one.
- Implement email add factor basic methods.

### CODE REVIEW

File: auth.php [hunk 1]
```diff
@@ -55,8 +64,17 @@
 
 if ($form->is_submitted()) {
     if ($data = $form->get_data()) {
-        $_SESSION['USER']->tool_mfa_authenticated = true;
-        redirect(new moodle_url($wantsurl));
+        $property = 'factor_'.$factorname.'_authenticated';
+        $_SESSION['USER']->$property = true;
```
Comment (brendanheywood) [inline auth.php]: please just use $SESSION or $USER

File: classes/local/factor/object_factor_base.php [hunk 2]
```diff
@@ -77,12 +77,22 @@ public function add_user_factor($data) {
...
-    public function get_user_factors($user) {
+    public function get_all_user_factors($user) {
+        return array();
+    }
+
+    public function get_enabled_user_factor($userid) {
+        $factors = $this->get_all_user_factors($userid);
+        foreach ($factors as $factor) {
+            if ($factor->disabled == 0) {
+                return $factor;
```
Comment (brendanheywood) [inline classes/local/factor/object_factor_base.php]: this doesn't look right, are you filtering this array?

File: classes/local/factor/object_factor.php [hunk 3]
```diff
@@ -36,6 +36,7 @@ public function define_add_factor_form_definition_after_data($mform);
     public function define_login_form($mform);
     public function validation($data);
     public function add_user_factor($data);
-    public function get_user_factors($user);
+    public function get_all_user_factors($user);
```
Comment (brendanheywood) [inline classes/local/factor/object_factor.php]: Can you please add php docs for all of these as you go so it's clear the intent of each of the methods. eg get_all_user_factors() isn't obvious to me what it's doing (yet)

File: classes/plugininfo/factor.php [hunk 4]
```diff
@@ -82,6 +82,48 @@ public static function get_enabled_factors() {
         return $return;
     }
 
+    /**
+     * Finds enabled factors for user by userid.
+     *
+     * @param int $userid user id.
+     * @return array of enabled factors for given user.
+     */
+    public static function get_enabled_user_factors($userid) {
```
Comment (brendanheywood) [inline classes/plugininfo/factor.php]: Is this supposed to return all factors for a user, or all factors for this particular type of factor for a particular user?

File: classes/local/form/login_form.php [hunk 5]
```diff
@@ -33,24 +33,40 @@ class login_form extends \moodleform
...
+        return $mform;
     }
 
     public function validation($data, $files) {
         $errors = parent::validation($data, $files);
 
-        $factors = \tool_mfa\plugininfo\factor::get_enabled_factors();
-        foreach ($factors as $factor) {
+        $factor = \tool_mfa\plugininfo\factor::get_factor($data['factor_name']);
+        if ($factor) {
```
Comment (brendanheywood) [inline classes/local/form/login_form.php]: if $factor is not valid lets make that an $error too

File: classes/plugininfo/factor.php [hunk 6]
```diff
@@ -82,6 +82,48 @@ public static function get_enabled_factors() {
...
+     */
+    public static function get_next_user_factor($userid) {
+        $factors = self::get_enabled_factors();
+
+        foreach ($factors as $factor) {
+            $userfactor = $factor->get_enabled_user_factor($userid);
+            if ($userfactor) {
+                $property = 'factor_'.$userfactor->name.'_authenticated';
+                if (empty($_SESSION['USER']->$property) || true !== $_SESSION['USER']->$property) {
+                    return $userfactor->name;
```
Comment (brendanheywood) [inline classes/plugininfo/factor.php]: why not just pass back the factor object instead?

File: factor/email/classes/factor.php [hunk 7]
```diff
@@ -31,4 +32,45 @@
 
 class factor extends object_factor_base {
 
+    public function define_add_factor_form_definition($mform) {
+        global $OUTPUT;
```
Comment (brendanheywood) [inline factor/email/classes/factor.php]: I'm confused about this form, we already know their email. Email should be a zero config factor as the user has already signed up with their email and confirmed it

File: factor/email/db/install.php [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+ * @package     factor_email
+ * @subpackage  tool_mfa
+ * @author      Mikhail Golenkov <golenkovm@gmail.com>
+ * @copyright   Catalyst IT
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+defined('MOODLE_INTERNAL') || die;
+
+function xmldb_factor_email_install() {
```
Comment (brendanheywood) [inline factor/email/db/install.php]: this file probably should not exist

File: factor/email/db/install.xml [hunk 9]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<XMLDB PATH="admin/tool/mfa/factor/email/db" VERSION="20191106" COMMENT="XMLDB file for Moodle tool/mfa plugin - email factor"
+       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xsi:noNamespaceSchemaLocation="../../../../lib/xmldb/xmldb.xsd"
+>
+    <TABLES>
+        <TABLE NAME="tool_mfa_factor_email" COMMENT="E-Mail Factor data for MFA plugin">
```
Comment (brendanheywood) [inline factor/email/db/install.xml]: See above, I don't think we need this table

File: factor/totp/db/install.xml [hunk 10]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8" ?>
-<XMLDB PATH="admin/tool/mfa/factor/totp/db" VERSION="20191104" COMMENT="XMLDB file for Moodle tool/mfa plugin"
+<XMLDB PATH="admin/tool/mfa/factor/totp/db" VERSION="20191104" COMMENT="XMLDB file for Moodle tool/mfa plugin - totp factor"
```
Comment (brendanheywood) [inline factor/totp/db/install.xml]: rename this table from tool_mfa_factor_totp to factor_totp

File: classes/local/factor/object_factor.php [hunk 11]
```diff
@@ -36,6 +36,7 @@ public function define_add_factor_form_definition_after_data($mform);
...
+    public function get_enabled_user_factor($user);
```
Comment (brendanheywood) [inline classes/local/factor/object_factor.php]: I think this should be passed a $user object, not a $userid

File: factor/email/classes/factor.php [hunk 12]
```diff
@@ -31,4 +32,71 @@
...
+        $disabled = (int)!$this->is_enabled();
+
+        $return = array();
+        $return[1] = new \stdClass();
+        $return[1]->id = $id;
+        $return[1]->name = $name;
+        $return[1]->useremail = $useremail;
+        $return[1]->timemodified = $timemodified;
+        $return[1]->timecreated = $timecreated;
+        $return[1]->disabled = $disabled;
```
Comment (brendanheywood) [inline factor/email/classes/factor.php]: This seems a bit redundant? why not just

```
$return = array(
  'id' => 1,
  'name' => $this->name
...
);
```

Also I would mix in $USER->confirmed as part of the enabled

File: factor/totp/classes/factor.php [hunk 13]
```diff
@@ -46,37 +46,19 @@
...
-                }
-            }
-        }
-        return array();
-    }
-
     public function draw_qrcode($secretcode) {
         global $USER;
-        $code = 'otpauth://totp/'.$USER->username.':'.$USER->email.'?secret='.$secretcode.'&issuer=Moodle&algorithm=SHA1&&period=30';
+        $code = 'otpauth://totp/'.$USER->username.'_2:'.$USER->email.'?secret='.$secretcode.'&issuer=Moodle&algorithm=SHA1&&period=30';
```
Comment (brendanheywood) [inline factor/totp/classes/factor.php]: typo?

File: factor/totp/classes/factor.php [hunk 14]
```diff
@@ -90,21 +72,63 @@ public function define_add_factor_form_definition($mform) {
...
-            if ($userfactor->disabled != 1) {
-                $mform->addElement('text', 'verificationcode', get_string('verificationcode', 'factor_totp'));
-                $mform->addRule('verificationcode', get_string('required'), 'required', null, 'client');
-                $mform->setType("verificationcode", PARAM_ALPHANUM);
-            }
+        //if (!empty($secretfield)) {
+        $secret = $secretfield->getValue();
+        $qrcode = $this->draw_qrcode($secret);
+
+        $mform->addElement('html', $OUTPUT->heading('Scan QR-Code or enter a key to your Google Authenticator:', 5));
```
Comment (brendanheywood) [inline factor/totp/classes/factor.php]: lang keys for these two headings

File: factor/totp/classes/factor.php [hunk 15]
```diff
@@ -90,21 +72,63 @@ public function define_add_factor_form_definition($mform) {
...
+        $factors = $this->get_enabled_user_factors();
+
+        foreach ($factors as $factor) {
+            if ($factor) {
+                $secret = $factor->secret;
+                $hotp = TOTP::create($secret);
+                $otp = $hotp->now();
+
+                if ($data['verificationcode'] != $otp) {
+                    return array('verificationcode' => 'Wrong verification code');
```
Comment (brendanheywood) [inline factor/totp/classes/factor.php]: won't this be wrong if you have multiple totp setup?

Shouldn't it default to an error, and as soon as you hit a match you remove the error?

File: factor/totp/classes/factor.php [hunk 16]
```diff
@@ -90,21 +72,63 @@ public function define_add_factor_form_definition($mform) {
...
         }
 
         return $mform;
     }
 
+    public function login_form_validation($data) {
+        $factors = $this->get_enabled_user_factors();
+
+        foreach ($factors as $factor) {
+            if ($factor) {
```
Comment (brendanheywood) [inline factor/totp/classes/factor.php]: is this if required?

File: factor/totp/classes/factor.php [hunk 17]
```diff
@@ -90,21 +72,63 @@ public function define_add_factor_form_definition($mform) {
...
+                if ($data['verificationcode'] != $otp) {
+                    return array('verificationcode' => 'Wrong verification code');
+                }
+            }
+        }
+        return array();
+    }
+
     public function get_secret_length() {
         $length = get_config('factor_totp', 'secret_length');
```
Comment (brendanheywood) [inline factor/totp/classes/factor.php]: I would actually remove this setting and hard code it to something very secure which is known to work everywhere. The less settings the better

File: lib.php [hunk 18]
```diff
@@ -24,7 +24,8 @@
 defined('MOODLE_INTERNAL') || die();
 
 function tool_mfa_after_require_login() {
-    if (empty($_SESSION['USER']->tool_mfa_authenticated) || true !== $_SESSION['USER']->tool_mfa_authenticated) {
+    global $USER;
+    if (empty($USER->tool_mfa_authenticated) || true !== $USER->tool_mfa_authenticated) {
         if ($GLOBALS['ME'] != '/admin/tool/mfa/auth.php') {
```
Comment (brendanheywood) [inline lib.php]: just use $ME same as $USER

File: auth.php (hunk 1 above)
Comment (golenkovm) [reply auth.php]: Corrected

File: classes/local/factor/object_factor_base.php (hunk 2 above)
Comment (golenkovm) [reply classes/local/factor/object_factor_base.php]: Corrected

File: classes/local/factor/object_factor.php (hunk 3 above)
Comment (golenkovm) [reply classes/local/factor/object_factor.php]: Will add php docs in further PR

File: classes/plugininfo/factor.php (hunk 4 above)
Comment (golenkovm) [reply classes/plugininfo/factor.php]: I agree that it's not a best choice for method name. Amended to make it more clear

File: classes/local/form/login_form.php (hunk 5 above)
Comment (golenkovm) [reply classes/local/form/login_form.php]: I re-wrote this block and implemented new parameter `grace_mode`, so now there is no need to check `if ($factor)`

File: classes/plugininfo/factor.php (hunk 6 above)
Comment (golenkovm) [reply classes/plugininfo/factor.php]: Corrected

File: factor/email/classes/factor.php (hunk 7 above)
Comment (golenkovm) [reply factor/email/classes/factor.php]: Improved email factor as discussed, so no there is no `add_factor_form` for email factor

File: factor/email/db/install.php (hunk 8 above)
Comment (golenkovm) [reply factor/email/db/install.php]: Corrected

File: factor/email/db/install.xml (hunk 9 above)
Comment (golenkovm) [reply factor/email/db/install.xml]: Corrected

File: factor/totp/db/install.xml (hunk 10 above)
Comment (golenkovm) [reply factor/totp/db/install.xml]: Corrected

File: classes/local/factor/object_factor.php (hunk 11 above)
Comment (golenkovm) [reply classes/local/factor/object_factor.php]: Removed argument at all. Using USER global variable in methods

File: factor/email/classes/factor.php (hunk 12 above)
Comment (golenkovm) [reply factor/email/classes/factor.php]: We need other fields to build a user preference table. Also we user `disabled` property in `get_enabled_user_factors()` method of factor base class

File: factor/totp/classes/factor.php (hunk 13 above)
Comment (golenkovm) [reply factor/totp/classes/factor.php]: In general, yes. Will discuss it later. Corrected

File: factor/totp/classes/factor.php (hunk 14 above)
Comment (golenkovm) [reply factor/totp/classes/factor.php]: Corrected

File: factor/totp/classes/factor.php (hunk 15 above)
Comment (golenkovm) [reply factor/totp/classes/factor.php]: You're right. Re-done this method, now it works fine

File: factor/totp/classes/factor.php (hunk 16 above)
Comment (golenkovm) [reply factor/totp/classes/factor.php]: Corrected

File: factor/totp/classes/factor.php (hunk 17 above)
Comment (golenkovm) [reply factor/totp/classes/factor.php]: Will be done in further PR

File: lib.php (hunk 18 above)
Comment (golenkovm) [reply lib.php]: Done

File: factor/email/classes/factor.php (hunk 12 above)
Comment (brendanheywood) [reply factor/email/classes/factor.php]: the first comment was purely a code formatting issue, it is 14 lines of code which could be 6 lines of code and do the same thing and be easier to read.

Comment (golenkovm) [reply factor/email/classes/factor.php]: It needs to be an array of objects, not an array of arrays.
However, it can be simplified. Thanks

Comment (brendanheywood) [reply factor/email/classes/factor.php]: You can cast arrays to objects like this:

```
$return = array((object) [
  'id' => 1,
  'name' => $this->name,
  'useremail' => $useremail,
   ...
]);
```

https://stackoverflow.com/questions/19272011/how-to-convert-an-array-into-an-object-using-stdclass/19272047

**Pull request merged**